### PR TITLE
[gh-2381] revise account sign command and add account verify command

### LINKED ISSUE #2381: [CLI]Need rooch sign and validator command
Rooch cli tools need sign and validator command. 

- sign       use account secret key to sign content hex bytes 
- validator  validator the signature by the rooch account signer.

### PR DESCRIPTION
**Pull request by feliciss**
## Summary

- Revise `account sign` command.
- Add `account verify` command.

- Closes #2381

### CODE REVIEW

File: crates/rooch/src/commands/account/commands/sign.rs [hunk 1]
```diff
@@ -1,61 +1,120 @@
...
+    ///
+    /// Supported types [u8, u16, u32, u64, u128, u256, bool, object_id, string, address, vector<inner_type>]
+    ///
+    /// Example: `address:0x1 bool:true u8:0 u256:1234 'vector<u32>:a,b,c,d'`
+    ///     address and uint can be written in short form like `@0x1 1u8 4123u256`.
+    #[clap(long = "args", value_parser=parse_function_arg)]
+    pub args: Vec<FunctionArg>,
+
+    #[clap(flatten)]
+    pub tx_options: TransactionOptions,
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/sign.rs:46]: We need a sign message command, not sign tx. the `tx sign` is for sign tx.

**Review by jolestar — CHANGES_REQUESTED** [review]

File: crates/rooch/src/commands/account/commands/sign.rs (hunk 1 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/sign.rs:46]: > We need a sign message command, not sign tx. the `tx sign` is for sign tx.

Noted.

**Comment by feliciss** [comment]
As discussed in https://discord.com/channels/1078938449974935592/1280474816121405510/1280822602008498239. I leave this PR unhandled.

File: crates/rooch/src/commands/account/commands/sign.rs [hunk 2]
```diff
@@ -5,57 +5,66 @@ use crate::cli_types::{CommandAction, WalletContextOptions};
...
-/// This operation must be specified with -a or
-/// --address to export only one address with a private key.
+/// Sign a message with the bitcoin address
 #[derive(Debug, Parser)]
 pub struct SignCommand {
-    #[clap(short = 'a', long = "address", value_parser=ParsedAddress::parse, default_value = "")]
-    address: ParsedAddress,
+    // the address to be used
+    #[clap(short = 'a', long = "btc-address")]
+    bitcoin_address: BitcoinAddress,
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/sign.rs:21]: Use the `ParsedAddress` to support all address

File: crates/rooch/src/commands/account/commands/sign.rs [hunk 3]
```diff
@@ -5,57 +5,66 @@ use crate::cli_types::{CommandAction, WalletContextOptions};
...
 
-        let signature =
-            context
-                .keystore
-                .sign_hashed(&addrss, &self.msg.clone().into_bytes(), password)?;
+        let auth_payload = AuthPayload::new_without_tx_hash(
+            sign_data,
+            signature,
+            self.bitcoin_address.to_string(),
+        );
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/sign.rs:58]: Do not need the auth_payload

File: crates/rooch/src/commands/account/commands/verify.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    error::{RoochError, RoochResult},
+    framework::auth_payload::AuthPayload,
+};
+
+/// Verify an auth payload hex
+#[derive(Debug, Parser)]
+pub struct VerifyCommand {
+    /// an auth payload hex for verify
+    #[clap(long)]
+    input: String,
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/verify.rs:17]: Input should be signature

File: crates/rooch/src/commands/account/commands/sign.rs (hunk 2 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/sign.rs]: > Use the `ParsedAddress` to support all address

Used.

File: crates/rooch/src/commands/account/commands/sign.rs (hunk 3 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/sign.rs]: > Do not need the auth_payload

Removed.

File: crates/rooch/src/commands/account/commands/verify.rs (hunk 4 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/verify.rs]: > Input should be signature

Added signature and message hash for verification.

File: crates/rooch/src/commands/account/commands/sign.rs [hunk 5]
```diff
@@ -5,57 +5,73 @@ use crate::cli_types::{CommandAction, WalletContextOptions};
...
         let password = context.get_password();
-
         let mapping = context.address_mapping();
-        let addrss = self.address.into_rooch_address(&mapping).map_err(|e| {
-            RoochError::CommandArgumentError(format!("Invalid Rooch address String: {}", e))
-        })?;
+        let rooch_address = self.address.into_rooch_address(&mapping)?;
+
+        let mut message_info = Vec::new();
+        message_info.append(&mut MESSAGE_INFO.to_vec());
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/sign.rs:52]: You do not need to include the MESSAGE_INFO const. The MESSAGE_INFO is the rooch transaction sign message, and this custom sign message is from the cli argument.

File: crates/rooch/src/commands/account/commands/verify.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+/// Verify a signature
+#[derive(Debug, Parser)]
+pub struct VerifyCommand {
+    /// A signature for verify
+    #[clap(short = 's', long)]
+    signature_hash: String,
+
+    /// A hashed message to be verified
+    #[clap(short = 'm', long)]
+    message_hash: String,
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/verify.rs:26]: The verify should accept the original message, not message hash.

File: crates/rooch/src/commands/account/commands/sign.rs [hunk 7]
```diff
@@ -5,57 +5,73 @@ use crate::cli_types::{CommandAction, WalletContextOptions};
...
+        let encoded_sign_data = sign_data.encode();
 
         let signature =
             context
                 .keystore
-                .sign_hashed(&addrss, &self.msg.clone().into_bytes(), password)?;
+                .sign_hashed(&rooch_address, &encoded_sign_data, password)?;
+
+        let output = SignAccountOutput {
+            signature_hash: signature.encode_hex(),
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/sign.rs:64]: This the signature, not signature_hash

File: crates/rooch/src/commands/account/commands/sign.rs [hunk 8]
```diff
@@ -5,57 +5,73 @@ use crate::cli_types::{CommandAction, WalletContextOptions};
...
+            message_hash: encoded_sign_data.encode_hex(),
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/sign.rs:65]: Do not need to return this.

File: crates/rooch/src/commands/account/commands/verify.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    crypto::{RoochSignature, Signature},
+    error::{RoochError, RoochResult},
+};
+
+/// Verify a signature
+#[derive(Debug, Parser)]
+pub struct VerifyCommand {
+    /// A signature for verify
+    #[clap(short = 's', long)]
+    signature_hash: String,
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/verify.rs:22]: signature not signature_hash

File: crates/rooch/src/commands/account/commands/verify.rs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            .map_err(|e| RoochError::CommandArgumentError(format!("Decode hex failed: {}", e)))?;
+        let signatrue = Signature::from_bytes(&signature_bytes).map_err(|e| {
+            RoochError::CommandArgumentError(format!("Invalid signature argument: {}", e))
+        })?;
+        let pk = Secp256k1PublicKey::from_bytes(signatrue.public_key_bytes())
+            .map_err(|e| RoochError::CommandArgumentError(format!("Invalid public key: {}", e)))?;
+        let sig = Secp256k1Signature::from_bytes(signatrue.signature_bytes()).map_err(|e| {
+            RoochError::CommandArgumentError(format!("Invalid signature argument: {}", e))
+        })?;
+        let message_hash = hex::decode(self.message_hash)
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/verify.rs:49]: We need to append the prefix to the message and calculate the message_hash again.

File: crates/testsuite/features/cmd.feature [hunk 11]
```diff
@@ -37,6 +37,10 @@ Feature: Rooch CLI integration tests
 
       Then cmd: "account create"
       Then cmd: "account list --json"
+      # account sign and verify
+      Then cmd: "account sign -a {{$.account[-1].account0.address}} -m 'empty' --json"
+      Then cmd: "account verify -s {{$.account[-1].signature_hash}} -m {{$.account[-1].message_hash}}"
```
Comment (jolestar) [inline crates/testsuite/features/cmd.feature:42]: Need to assert the result is ok.

File: crates/rooch/src/commands/account/commands/sign.rs (hunk 5 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/sign.rs]: > You do not need to include the MESSAGE_INFO const. The MESSAGE_INFO is the rooch transaction sign message, and this custom sign message is from the cli argument.

Resolved.

File: crates/rooch/src/commands/account/commands/verify.rs (hunk 6 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/verify.rs]: Resolved.

File: crates/rooch/src/commands/account/commands/sign.rs (hunk 7 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/sign.rs]: Resolved. But there needs a conversion from output Signature to input Signature string.

File: crates/rooch/src/commands/account/commands/sign.rs (hunk 8 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/sign.rs]: Removed.

File: crates/rooch/src/commands/account/commands/verify.rs (hunk 9 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/verify.rs]: There needs a value parser to convert them from output to the input.

File: crates/rooch/src/commands/account/commands/verify.rs (hunk 10 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/verify.rs]: > We need to append the prefix to the message and calculate the message_hash again.

Resolved.

File: crates/testsuite/features/cmd.feature (hunk 11 above)
Comment (feliciss) [reply crates/testsuite/features/cmd.feature]: Asserted.

**Review by wow-sven — APPROVED** [review]

File: crates/rooch/src/commands/account/commands/verify.rs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+#[async_trait]
+impl CommandAction<Option<bool>> for VerifyCommand {
+    async fn execute(self) -> RoochResult<Option<bool>> {
+        let sign_data =
+            SignData::new_without_tx_hash(MESSAGE_INFO_PREFIX.to_vec(), self.message.to_bytes());
+        let encoded_sign_data = sign_data.encode();
+        let message_hash = sha2_256_of(&encoded_sign_data).0.to_vec();
+        let _ = self.signature.into_inner().verify(&message_hash);
```
Comment (jolestar) [inline crates/rooch/src/commands/account/commands/verify.rs:41]: The sign result should not be ignored.

**Review by jolestar — CHANGES_REQUESTED** [review]

File: crates/rooch/src/commands/account/commands/verify.rs (hunk 12 above)
Comment (feliciss) [reply crates/rooch/src/commands/account/commands/verify.rs]: > The sign result should not be ignored.

Removed the `sha_256_of` and fixed the issue.

**Review by jolestar — APPROVED** [review]

**Pull request merged**